File: src/main/java/com/hileco/cortex/primitives/layer/Layer.java
package com.hileco.cortex.primitives.layer;

import java.util.ArrayList;
import java.util.List;

public abstract class Layer<T extends Layer<T>> {
    private final List<T> children;
    private T parent;

    public Layer(T parent) {
        this.children = new ArrayList<>();
        this.parent = parent;
        while (this.parent != null && this.parent.isEmpty()) {
            this.parent = this.parent.getParent();
        }
        if (this.parent != null) {
            this.parent.getChildren().add((T) this);
        }
    }

    public List<T> getChildren() {
        return children;
    }

    public T getParent() {
        return parent;
    }

    abstract boolean isEmpty();

    public void close() {
        var currentParent = this.parent;
        while (currentParent != null) {
            currentParent.getChildren().remove(this);
            if (currentParent.getChildren().isEmpty()) {
                currentParent = currentParent.getParent();
            }
        }
        this.parent = null;
    }
}

File: cortex-collections/src/main/kotlin/com/hileco/cortex/collections/layer/Layer.kt
package com.hileco.cortex.collections.layer

import java.util.*

/**
 * [Layer] structures' internal layering must be thread-safe.
 *
 * Objects implementing [Layer] do not have to be thread safe
 */
@Suppress("UNCHECKED_CAST")
abstract class Layer<T : Layer<T>>(parent: T?) {
    val children: MutableList<T>
    var parent: T? = null

    internal abstract val isEmpty: Boolean

    init {
        this.children = ArrayList()
        var currentParent = parent
        while (currentParent != null && currentParent.isEmpty) {
            currentParent = currentParent.parent
        }
        this.parent = currentParent
        currentParent?.addChild(this as T)
    }

    @Synchronized
    private fun addChild(child: T) {
        this.children.add(child)
    }

    @Synchronized
    private fun removeChild(child: T): Boolean {
        this.children.remove(child)
        return this.children.isEmpty()
    }

    fun close() {
        var currentParent = this.parent
        while (currentParent != null) {
            val removedLastChild = currentParent.removeChild(this as T)
            if (removedLastChild) {
                currentParent = currentParent.parent
            } else {
                break
            }
        }
        this.parent = null
    }
}
